bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py

```diff
@@ -0,0 +1,278 @@
+
+        # Define signature
+        pywrapper_signature = [
+                'ptr_{0}, size_{0}'.format(i) if 'RVec' in t else 'x_{}'.format(i) \
+                        for i, t in enumerate(input_types)]
+        if 'RVec' in return_type:
+            # If we return an RVec, we return via pointer the pointer of the allocated data,
+            # the size in elements. In addition, we provide the size of the datatype in bytes.
+            pywrapper_signature += ['ptrptr_r, ptrsize_r', 'dtypesize_r']
+
+        # Define arguments for jit function
+        pywrapper_args_def = [
+                'x_{0} = nb.carray(ptr_{0}, (size_{0},))'.format(i) if 'RVec' in t else 'x_{}'.format(i) \
+                        for i, t in enumerate(input_types)]
+        pywrapper_args = ['x_{}'.format(i) for i in range(len(input_types))]
+
+        # Define return operation
+        if 'RVec' in return_type:
+            pywrapper_return = '\n    '.join([
+                '# Because we cannot manipulate the memory management of the numpy array we copy the data',
+                'ptr = malloc(r.size * dtypesize_r)',
+                'cp = nb.carray(ptr, r.size, dtype_r)',
+                'cp[:] = r[:]',
+                '# Return size of the array and the pointer to the copied data',
+                'ptrsize_r[0] = r.size',
+                'ptrptr_r[0] = cp.ctypes.data'
+                ])
+        else:
+            pywrapper_return = 'return r'
+
+        # Build wrapper code
+        pywrappercode = '''\
+def pywrapper({SIGNATURE}):
+    """
+    Wrapper function for the jitted Python callable with special treatment of arrays
+    """
+    # If an RVec is given, define numba carray wrapper for the input types
+    {ARGS_DEF}
+    # Call the jitted Python function
+    r = nbjit({ARGS})
+    # Return the result
+    {RETURN}
+        '''.format(
+                SIGNATURE=', '.join(pywrapper_signature),
+                ARGS_DEF='\n    '.join(pywrapper_args_def),
+                ARGS=', '.join(pywrapper_args),
+                RETURN=pywrapper_return
+                )
+
+        glob = dict(globals()) # Make a shallow copy of the dictionary so we don't pollute the global scope
+        glob['nb'] = nb
+        glob['nbjit'] = nbjit
+
+        ffi = cffi.FFI()
+        ffi.cdef('void* malloc(long size);')
+        C = ffi.dlopen(None)
+        glob['malloc'] = C.malloc
+
+        if 'RVec' in return_type:
+            glob['dtype_r'] = get_numba_type(get_inner_type(return_type))
+
+        exec(pywrappercode, glob, locals())
+        if not 'pywrapper' in locals():
+            raise Exception('Failed to create Python wrapper function:\n{}'.format(pywrappercode))
+
+        # Jit the Python wrapper code
+        c_return_type, c_input_types = get_c_signature(input_types, return_type)
+        try:
+            nbcfunc = nb.cfunc(c_return_type(*c_input_types), nopython=True, inline='always')(locals()['pywrapper'])
```

Comment:
i don't think it actually hurts, but we don't really need the `inline='always'` here

Reply:
True, shouldn't change anything. I left it there for completeness.

Reply:
it's a bit of extra mental load for the reader because now you are thinking you did not understand the logic because you don't expect this function to be inlined in another numba function

Reply:
removed